draft-ietf-quic-transport.md

```diff
@@ -2307,10 +2307,14 @@ signal closure.
 If the connection has been successfully established, endpoints MUST send any
 CONNECTION_CLOSE frames in a 1-RTT packet.  Prior to connection establishment a
 peer might not have 1-RTT keys, so endpoints SHOULD send CONNECTION_CLOSE frames
-in a Handshake packet.  If the endpoint does not have Handshake keys, or it is
-not certain that the peer has Handshake keys, it MAY send CONNECTION_CLOSE
-frames in an Initial packet.  If multiple packets are sent, they can be
-coalesced (see {{packet-coalesce}}) to facilitate retransmission.
+in a Handshake packet.  If the endpoint does not have Handshake keys, it SHOULD
+send CONNECTION_CLOSE frames in an Initial packet.
+
+The server may not know whether the client has Handshake keys.  In order to
+guarantee a CONNECTION_CLOSE is processed, it SHOULD send a CONNECTION_CLOSE
+in both Handshake and Initial, because the client discards Initial keys as soon
+as it has Handshake keys. If multiple packets are sent, they can be coalesced
+(see {{packet-coalesce}}).
```

Comment:
Suggested rewording of text after the first sentence:
"Under these circumstances, an endpoint SHOULD send a CONNECTION_CLOSE frame in both Handshake and Initial packets to increase the likelihood of a CONNECTION_CLOSE frame being processed by the peer.  These packets can be coalesced into a single UDP datagram (see {{packet-coalesce}})."

Reply:
A bigger point here is that the server also isn't sure when the client gets 1-RTT keys.  We should say the same about that transition.  I believe that we can rely on handshake confirmation for this transition.  Endpoints might choose to send CONNECTION_CLOSE in Handshake packets prior to the handshake being confirmed.

Reply:
Agreed, another spot we're relying on #2673?

Reply:
I updated the text about sending 1-RTT connection closes to use Handshake confirmed, which lines up with #2673.  I think that's a good fix?

Reply:
Don't sound so uncertain :)  I think that the fix is good.

And we'd (collectively) be crazy if we didn't take #2673.

Reply:
We'd also individually go crazy if we don't take it.  😉

Reply:
"Under these circumstances, a server SHOULD send a CONNECTION_CLOSE frame in both Handshake and Initial packets to ensure that at least one of them is processable by the client. These packets can be coalesced into a single UDP datagram (see {{packet-coalesce}})."